diff --git a/Coordinator.cpp b/Coordinator.cpp
index ab693f68..5e2e70e4 100644
--- a/Coordinator.cpp
+++ b/Coordinator.cpp
@@ -31,6 +31,10 @@
 #include "Interface.h"
 #include "hidl-gen_l.h"
 
+#if defined(__linux__)
+#include <cstring>
+#endif
+
 static bool existdir(const char *name) {
     DIR *dir = opendir(name);
     if (dir == nullptr) {
@@ -49,9 +53,15 @@ const std::string &Coordinator::getRootPath() const {
 void Coordinator::setRootPath(const std::string &rootPath) {
     mRootPath = rootPath;
 
+#if defined(_MSC_VER)
+    if (!mRootPath.empty() && !StringHelper::EndsWith(mRootPath, "\\")) {
+        mRootPath += "\\";
+    }
+#else
     if (!mRootPath.empty() && !StringHelper::EndsWith(mRootPath, "/")) {
         mRootPath += "/";
     }
+#endif
 }
 
 void Coordinator::setOutputPath(const std::string& outputPath) {
@@ -390,7 +400,11 @@ const Coordinator::PackageRoot* Coordinator::findPackageRoot(const FQName& fqNam
 }
 
 std::string Coordinator::makeAbsolute(const std::string& path) const {
+#if defined(_MSC_VER)
+    if (StringHelper::StartsWith(path, "\\") || mRootPath.empty()) {
+#else
     if (StringHelper::StartsWith(path, "/") || mRootPath.empty()) {
+#endif
         return path;
     }
 
@@ -430,12 +444,20 @@ status_t Coordinator::getPackagePath(const FQName& fqName, bool relative, bool s
 
     std::vector<std::string> components;
     if (!relative) {
+#if defined(_MSC_VER)
+        components.push_back(StringHelper::RTrimAll(packageRoot->path, "\\"));
+#else
         components.push_back(StringHelper::RTrimAll(packageRoot->path, "/"));
+#endif
     }
     components.insert(components.end(), suffixComponents.begin(), suffixComponents.end());
     components.push_back(sanitized ? fqName.sanitizedVersion() : fqName.version());
 
+#if defined(_MSC_VER)
+    *path = StringHelper::JoinStrings(components, "\\") + "\\";
+#else
     *path = StringHelper::JoinStrings(components, "/") + "/";
+#endif
     return OK;
 }
 
@@ -530,11 +552,20 @@ status_t Coordinator::convertPackageRootToPath(const FQName& fqName, std::string
     status_t err = getPackageRoot(fqName, &packageRoot);
     if (err != OK) return err;
 
+#if defined(_MSC_VER)
+    std::string::const_iterator end_itr = (packageRoot.end()-1);
+    if (*end_itr != '.') {
+#else
     if (*(packageRoot.end()--) != '.') {
+#endif
         packageRoot += '.';
     }
 
+#if defined(_MSC_VER)
+    std::replace(packageRoot.begin(), packageRoot.end(), '.', '\\');
+#else
     std::replace(packageRoot.begin(), packageRoot.end(), '.', '/');
+#endif
 
     *path = packageRoot;  // now converted to a path
     return OK;
@@ -920,7 +951,11 @@ bool Coordinator::MakeParentHierarchy(const std::string &path) {
 
     size_t start = 1;  // Ignore leading '/'
     size_t slashPos;
+#if defined(_MSC_VER) 
+    while ((slashPos = path.find('\\', start)) != std::string::npos) {
+#else
     while ((slashPos = path.find('/', start)) != std::string::npos) {
+#endif
         std::string partial = path.substr(0, slashPos);
 
         struct stat st;
@@ -929,7 +964,11 @@ bool Coordinator::MakeParentHierarchy(const std::string &path) {
                 return false;
             }
 
+#if defined(_MSC_VER)
+            int res = mkdir(partial.c_str());
+#else
             int res = mkdir(partial.c_str(), kMode);
+#endif
             if (res < 0) {
                 return false;
             }
diff --git a/EnumType.cpp b/EnumType.cpp
index 32682645..9f49e4f5 100644
--- a/EnumType.cpp
+++ b/EnumType.cpp
@@ -25,6 +25,10 @@
 #include "Location.h"
 #include "ScalarType.h"
 
+#if defined(__linux__)
+#include <algorithm>
+#endif
+
 namespace android {
 
 EnumType::EnumType(const char* localName, const FQName& fullName, const Location& location,
@@ -246,6 +250,18 @@ void EnumType::emitTypeDeclarations(Formatter& out) const {
     const ScalarType *scalarType = mStorageType->resolveToScalarType();
     CHECK(scalarType != nullptr);
 
+#if defined(_MSC_VER)
+    if (localName() == "Status") {
+        out << "#if defined(NO_ERROR)\n"
+            << "#undef NO_ERROR\n"
+            << "#endif\n";
+    } else if (localName() == "Transport") {
+        out << "#if defined(PASSTHROUGH)\n"
+            << "#undef PASSTHROUGH\n"
+            << "#endif\n";
+    } 
+#endif
+
     const std::string storageType = scalarType->getCppStackType();
 
     out << "enum class "
diff --git a/HidlTypeAssertion.cpp b/HidlTypeAssertion.cpp
index 6e4822ad..22d7347b 100644
--- a/HidlTypeAssertion.cpp
+++ b/HidlTypeAssertion.cpp
@@ -21,6 +21,10 @@
 #include <string>
 #include <vector>
 
+#if defined(_MSC_VER) || defined(__linux__)
+#include <algorithm>
+#endif
+
 namespace android {
 
 typedef std::vector<std::pair<std::string, size_t>> Registry;
diff --git a/Interface.cpp b/Interface.cpp
index b0ffa34d..fd22f841 100644
--- a/Interface.cpp
+++ b/Interface.cpp
@@ -321,12 +321,14 @@ bool Interface::fillHashChainMethod(Method *method) const {
         HIDL_HASH_CHAIN_TRANSACTION,
         { { IMPL_INTERFACE, [this, digestType](auto &out) {
             std::vector<const Interface *> chain = typeChain();
+#if !defined(_MSC_VER)  // M3E: the MSVC doesn't support it
             out << "_hidl_cb(";
             out.block([&] {
                 emitDigestChain(out, "(" + digestType->getInternalDataCppType() + ")", chain,
                                 [](const auto& e) { return e->cppValue(); });
             });
             out << ");\n";
+#endif
             out << "return ::android::hardware::Void();\n";
         } } }, /* cppImpl */
         { { IMPL_INTERFACE, [this, digestType, chainType](auto &out) {
diff --git a/TypeDef.cpp b/TypeDef.cpp
index 2543d49a..4c874385 100644
--- a/TypeDef.cpp
+++ b/TypeDef.cpp
@@ -74,6 +74,13 @@ bool TypeDef::resultNeedsDeref() const {
 }
 
 void TypeDef::emitTypeDeclarations(Formatter& out) const {
+#if defined(__linux__)
+    if (localName() == "Status") {
+        out << "#if defined(Status) \n"
+            << "#undef Status\n"
+            << "#endif\n";
+    }
+#endif
     out << "typedef "
         << mReferencedType->getCppStackType()
         << " "
diff --git a/generateCpp.cpp b/generateCpp.cpp
index 06110ca1..c7cab684 100644
--- a/generateCpp.cpp
+++ b/generateCpp.cpp
@@ -261,8 +261,16 @@ void AST::generateInterfaceHeader(Formatter& out) const {
     if (iface) {
         iface->emitDocComment(out);
 
+#if defined(_MSC_VER) // M3E: MSVC export
+        std::vector<std::string> nameComponents;
+        StringHelper::SplitString(mPackage.package(), '.', &nameComponents);
+        std::string component = StringHelper::Uppercase(nameComponents[1]);
+
+        out << "struct EXPORT_" << component << "  " << ifaceName;
+#else
         out << "struct "
             << ifaceName;
+#endif
 
         const Interface *superType = iface->superType();
 
@@ -855,7 +863,9 @@ void AST::generateCppSource(Formatter& out) const {
 
         generateCppPackageInclude(out, mPackage, iface->getProxyName());
         generateCppPackageInclude(out, mPackage, iface->getStubName());
+#if 0 // M3E: it is using auto which is not supported
         generateCppPackageInclude(out, mPackage, iface->getPassthroughName());
+#endif
 
         for (const Interface *superType : iface->superTypeChain()) {
             generateCppPackageInclude(out,
@@ -874,7 +884,9 @@ void AST::generateCppSource(Formatter& out) const {
     enterLeaveNamespace(out, true /* enter */);
     out << "\n";
 
+#if 0 // M3E:
     generateTypeSource(out, iface ? iface->localName() : "");
+#endif
 
     if (iface) {
         const Interface* iface = mRootScope.getInterface();
@@ -885,6 +897,7 @@ void AST::generateCppSource(Formatter& out) const {
             << "::descriptor(\""
             << iface->fqName().string()
             << "\");\n\n";
+#if 0 // M3E:
         out << "__attribute__((constructor)) ";
         out << "static void static_constructor() {\n";
         out.indent([&] {
@@ -931,11 +944,14 @@ void AST::generateCppSource(Formatter& out) const {
                 << "::descriptor);\n";
         });
         out << "};\n\n";
+#endif
 
         generateInterfaceSource(out);
+#if 0 // M3E:
         generateProxySource(out, iface->fqName());
         generateStubSource(out, iface);
         generatePassthroughSource(out);
+#endif
 
         if (isIBase()) {
             out << "// skipped getService, registerAsService, registerForNotifications\n";
@@ -947,7 +963,9 @@ void AST::generateCppSource(Formatter& out) const {
         }
     }
 
+#if 0 // M3E:
     HidlTypeAssertion::EmitAll(out);
+#endif
     out << "\n";
 
     enterLeaveNamespace(out, false /* enter */);
diff --git a/hidl-gen_y.yy b/hidl-gen_y.yy
index 65f43232..2fb02045 100644
--- a/hidl-gen_y.yy
+++ b/hidl-gen_y.yy
@@ -38,6 +38,10 @@
 #include <hidl-util/StringHelper.h>
 #include <stdio.h>
 
+#if defined(__linux__) // M3E:
+#include <algorithm>
+#endif
+
 using namespace android;
 
 extern int yylex(yy::parser::semantic_type*, yy::parser::location_type*, void*, Scope** const);
diff --git a/main.cpp b/main.cpp
index c03b6fcd..b67b5dc7 100644
--- a/main.cpp
+++ b/main.cpp
@@ -32,6 +32,14 @@
 #include <string>
 #include <vector>
 
+#if defined(_MSC_VER) || defined(__linux__) // M3E:
+#include <algorithm>
+#include <getopt.h>
+#endif
+#if defined(__linux__)
+#include <cstring>
+#endif
+
 using namespace android;
 
 enum class OutputMode {
@@ -1388,9 +1396,15 @@ int main(int argc, char **argv) {
             }
 
             if (outputFormat->mOutputMode == OutputMode::NEEDS_DIR) {
+#if defined(_MSC_VER)
+                if (outputPath.back() != '\\') {
+                    outputPath += "\\";
+                }
+#else
                 if (outputPath.back() != '/') {
                     outputPath += "/";
                 }
+#endif
             }
             break;
         }
@@ -1398,10 +1412,15 @@ int main(int argc, char **argv) {
             if (outputPath.empty()) {
                 outputPath = coordinator.getRootPath();
             }
+#if defined(_MSC_VER)
+            if (outputPath.back() != '\\') {
+                outputPath += "\\";
+            }
+#else
             if (outputPath.back() != '/') {
                 outputPath += "/";
             }
-
+#endif
             break;
         }
 
@@ -1413,10 +1432,17 @@ int main(int argc, char **argv) {
     coordinator.setOutputPath(outputPath);
 
     if (!suppressDefaultPackagePaths) {
+#if defined(_MSC_VER)
+        coordinator.addDefaultPackagePath("android.hardware", "hardware\\interfaces");
+        coordinator.addDefaultPackagePath("android.hidl", "system\\libhidl\\transport");
+        coordinator.addDefaultPackagePath("android.frameworks", "frameworks\\hardware\\interfaces");
+        coordinator.addDefaultPackagePath("android.system", "system\\hardware\\interfaces");
+#else
         coordinator.addDefaultPackagePath("android.hardware", "hardware/interfaces");
         coordinator.addDefaultPackagePath("android.hidl", "system/libhidl/transport");
         coordinator.addDefaultPackagePath("android.frameworks", "frameworks/hardware/interfaces");
         coordinator.addDefaultPackagePath("android.system", "system/hardware/interfaces");
+#endif
     }
 
     for (int i = 0; i < argc; ++i) {
