diff --git a/source/dng_mutex.cpp b/source/dng_mutex.cpp
index a8b2a6f..289fa5b 100644
--- a/source/dng_mutex.cpp
+++ b/source/dng_mutex.cpp
@@ -332,8 +332,12 @@ bool dng_condition::Wait (dng_mutex &mutex, double timeoutSecs)
 		
 	else
 		{
-		
+
+#if defined(_MSC_VER)		
+		struct dng_timespec now;
+#else
 		struct timespec now;
+#endif
 
 		dng_pthread_now (&now);
 
diff --git a/source/dng_pthread.cpp b/source/dng_pthread.cpp
index 0347856..b49206c 100644
--- a/source/dng_pthread.cpp
+++ b/source/dng_pthread.cpp
@@ -662,9 +662,12 @@ int dng_pthread_cond_wait(dng_pthread_cond_t *cond, dng_pthread_mutex_t *mutex)
 int dng_pthread_cond_timedwait(dng_pthread_cond_t *cond, dng_pthread_mutex_t *mutex, struct dng_timespec *latest_time)
 {
 	ValidateCond(cond);
-	
+
+#if defined(_MSC_VER)
 	struct dng_timespec sys_timespec;
-	
+#else
+	struct timespec sys_timespec;
+#endif
 	dng_pthread_now (&sys_timespec);
 
 	__int64 sys_time  = (__int64)sys_timespec.tv_sec * 1000000000 + sys_timespec.tv_nsec;
@@ -1096,7 +1099,11 @@ void dng_pthread_terminate()
 
 /*****************************************************************************/
 
+#if defined(_MSC_VER)
+int dng_pthread_now (struct dng_timespec *now)
+#else
 int dng_pthread_now (struct timespec *now)
+#endif
 	{
 	
 	if (now == NULL)
diff --git a/source/dng_pthread.h b/source/dng_pthread.h
index 0a28a1b..961444f 100644
--- a/source/dng_pthread.h
+++ b/source/dng_pthread.h
@@ -179,7 +179,9 @@ void dng_pthread_terminate();
 #undef PTHREAD_ONCE_INIT
 #define PTHREAD_ONCE_INIT DNG_PTHREAD_ONCE_INIT
 
+#if !defined(_MSC_VER)
 #define timespec dng_timespec
+#endif
 
 /* If it is defined on Windows, it probably has the wrong value... */
 #if defined(WIN32) || !defined(ETIMEDOUT)
@@ -249,7 +251,11 @@ extern "C"
 {
 #endif
 
+#if defined(_MSC_VER)
+int dng_pthread_now (struct dng_timespec *now);
+#else
 int dng_pthread_now (struct timespec *now);
+#endif
 
 #ifdef __cplusplus
 }
