diff --git a/src/common/third_party/base/anglebase/logging.h b/src/common/third_party/base/anglebase/logging.h
index 73f81e87f..9e3e022bc 100644
--- a/src/common/third_party/base/anglebase/logging.h
+++ b/src/common/third_party/base/anglebase/logging.h
@@ -20,7 +20,7 @@
 
 // Unfortunately ANGLE relies on ASSERT being an empty statement, which these libs don't respect.
 #ifndef NOTREACHED
-#    define NOTREACHED() ({ UNREACHABLE(); })
+#    define NOTREACHED() //({ UNREACHABLE(); })
 #endif
 
 #endif  // ANGLEBASE_LOGGING_H_
diff --git a/src/common/third_party/base/anglebase/numerics/safe_conversions.h b/src/common/third_party/base/anglebase/numerics/safe_conversions.h
index b37a36cd3..8bbb073ec 100644
--- a/src/common/third_party/base/anglebase/numerics/safe_conversions.h
+++ b/src/common/third_party/base/anglebase/numerics/safe_conversions.h
@@ -90,7 +90,7 @@ constexpr Dst saturated_cast_impl(const Src value, const RangeConstraint constra
                              ? std::numeric_limits<Dst>::max()
                              : (constraint == RANGE_INVALID
                                     ? NaNHandler::template HandleNaN<Dst>()
-                                    : (NOTREACHED(), static_cast<Dst>(value)))));
+                                    : (/*NOTREACHED(),*/ static_cast<Dst>(value)))));
 }
 }  // namespace internal
 
