diff --git a/include/GLFW/glfw3native.h b/include/GLFW/glfw3native.h
index 267e75ca..746320c5 100644
--- a/include/GLFW/glfw3native.h
+++ b/include/GLFW/glfw3native.h
@@ -215,6 +215,9 @@ GLFWAPI CGDirectDisplayID glfwGetCocoaMonitor(GLFWmonitor* monitor);
  *  @ingroup native
  */
 GLFWAPI id glfwGetCocoaWindow(GLFWwindow* window);
+
+// M3E: Add
+GLFWAPI id glfwGetCocoaLayer(GLFWwindow* window);
 #endif
 
 #if defined(GLFW_EXPOSE_NATIVE_NSGL)
diff --git a/src/cocoa_window.m b/src/cocoa_window.m
index 129e975e..12f35c41 100644
--- a/src/cocoa_window.m
+++ b/src/cocoa_window.m
@@ -835,6 +835,9 @@ static GLFWbool createNativeWindow(_GLFWwindow* window,
     window->ns.view = [[GLFWContentView alloc] initWithGlfwWindow:window];
     window->ns.retina = wndconfig->ns.retina;
 
+    // M3E: for ANGLE
+    [window->ns.view setWantsLayer:YES];
+
     if (fbconfig->transparent)
     {
         [window->ns.object setOpaque:NO];
@@ -1818,3 +1821,10 @@ GLFWAPI id glfwGetCocoaWindow(GLFWwindow* handle)
     return window->ns.object;
 }
 
+// M3E: Add
+GLFWAPI id glfwGetCocoaLayer(GLFWwindow* handle)
+{
+    _GLFWwindow* window = (_GLFWwindow*) handle;
+    _GLFW_REQUIRE_INIT_OR_RETURN(nil);
+    return [window->ns.view layer];
+}
