diff --git a/include/config/SkUserConfig.h b/include/config/SkUserConfig.h
index b8022043f9..8441a03ae2 100644
--- a/include/config/SkUserConfig.h
+++ b/include/config/SkUserConfig.h
@@ -9,13 +9,13 @@
   #define SK_GAMMA_APPLY_TO_A8
   #define SK_GAMMA_CONTRAST 0.0
   #define SK_GAMMA_EXPONENT 1.4
-  #define SK_HAS_HEIF_LIBRARY
+  //#define SK_HAS_HEIF_LIBRARY
   #define SK_HAS_JPEG_LIBRARY
   #define SK_HAS_PNG_LIBRARY
-  #define SK_HAS_WEBP_LIBRARY
-  #define SK_PDF_USE_SFNTLY
+  //#define SK_HAS_WEBP_LIBRARY
+  //#define SK_PDF_USE_SFNTLY
   #define SK_SUPPORT_PDF
-  #define SK_VULKAN
+  //#define SK_VULKAN
   #define SK_VULKAN_HEADER "Skia_Vulkan_Android.h"
   #define SK_XML
 #endif//SkUserConfig_DEFINED
diff --git a/include/config/SkUserConfigManual.h b/include/config/SkUserConfigManual.h
index 69b2d17799..646925f298 100644
--- a/include/config/SkUserConfigManual.h
+++ b/include/config/SkUserConfigManual.h
@@ -9,7 +9,7 @@
 #define SkUserConfigManual_DEFINED
   #define GR_GL_CUSTOM_SETUP_HEADER "gl/GrGLConfig_chrome.h"
   #define GR_TEST_UTILS 1
-  #define SK_BUILD_FOR_ANDROID_FRAMEWORK
+  //#define SK_BUILD_FOR_ANDROID_FRAMEWORK
   #define SK_DEFAULT_FONT_CACHE_LIMIT   (768 * 1024)
   #define SK_DEFAULT_GLOBAL_DISCARDABLE_MEMORY_POOL_SIZE (512 * 1024)
   #define SK_USE_FREETYPE_EMBOLDEN
