diff --git a/freeglut/freeglut/src/fg_internal.h b/freeglut/freeglut/src/fg_internal.h
index f487ece..74a60bd 100644
--- a/freeglut/freeglut/src/fg_internal.h
+++ b/freeglut/freeglut/src/fg_internal.h
@@ -61,7 +61,7 @@
 
 #elif defined(__APPLE__)
 /* This is a placeholder until we get native OSX support ironed out -- JFF 11/18/09 */
-#   define  TARGET_HOST_POSIX_X11  1
+#define TARGET_MAC_OSX  1
 /* #   define  TARGET_HOST_MAC_OSX    1 */
 
 #else
@@ -210,6 +210,9 @@
 #if TARGET_HOST_BLACKBERRY
 #include "blackberry/fg_internal_blackberry.h"
 #endif
+#if TARGET_MAC_OSX
+#include "osx/fg_internal_osx.h"
+#endif
 
 
 /* -- GLOBAL TYPE DEFINITIONS ---------------------------------------------- */
